src/main/src/main/java/org/geoserver/catalog/LayerGroupHelper.java

```diff
@@ -280,25 +322,101 @@ public String getLoopAsString(Stack<LayerGroupInfo> path) {
         }        
         return s.toString();
     }
-    
+
+    /**
+     * Check if a layer group contains recursive structures
+     *
+     * @param group The current layer group
+     * @param path Stack of each visited/parent LayerGroup
+     * @return true if the LayerGroup contains itself, or another LayerGroup contains itself
+     */
     private static boolean checkLoops(LayerGroupInfo group, Stack<LayerGroupInfo> path) {
         path.push(group);
         if (group.getLayers() != null) {
-            for (PublishedInfo child : group.getLayers()) {
+            int size = group.getLayers().size();
+            for (int i = 0; i < size; i++) {
+                PublishedInfo child = group.getLayers().get(i);
+                StyleInfo s;
+                //Handle incomplete layer groups, especially those constructed by the XStreamPersister
+                if (group.getStyles() == null || group.getStyles().size() == 0) {
+                    s = null;
+                } else {
+                    s = group.getStyles().get(i);
+                }
                 if (child instanceof LayerGroupInfo) {
                     if (isGroupInStack((LayerGroupInfo) child, path)) {
                         path.push((LayerGroupInfo) child);
                         return true;
                     } else if (checkLoops((LayerGroupInfo) child, path)) {
                         return true;
-                    }                
+                    }
+                } else if (child == null && s != null) {
+                    if (checkStyleGroupLoops(s, group, path)) {
+                        return true;
+                    }
                 }
             }
         }
         
         path.pop();
         return false;
     }
+
+    /**
+     * Check if a style group contains the enclosing layer group, or other recursive structures.
+     *
+     * @param styleGroup The style group
+     * @param group The current layer group
+     * @param path Stack of each visited/parent LayerGroup
+     * @return true if the style group contains itself, or another LayerGroup contains itself
+     */
+    private static boolean checkStyleGroupLoops(StyleInfo styleGroup, LayerGroupInfo group, Stack<LayerGroupInfo> path) {
+        try {
+            Catalog catalog = getCatalogFromStyle(styleGroup);
+            StyledLayerDescriptor sld = new GeoServerDataDirectory(catalog.getResourceLoader()).parsedSld(styleGroup);
```

Comment:
If you are just using the catalog here to make a flyweight GeoServerDataDirectory wrapper around a resource loader .... can I suggest making a StyleInfo.getSLD() method to avoid all this violation of encapsulation (and make everybody more testable).

Reply:
I would much prefer modifying StyleInfo, so I will do that. Was initially a bit hesitant to make unplanned API changes to _Info classes.